src/gencost/waterfall.py

```diff
@@ -2256,3 +2267,480 @@ def valid_generation(df_):
+        for
+
+        process:
+        1) instances when a generator doesn't report for entire patio range
+        2) instances when a single fuel generator switches fuel
+        3) zeroes reported (tbd)
+
+        output: df with plant / gen / year / prime / fuel / ba code / age / fuss
+        that we want to fill in with similar plants
+
+        """
+        hist_data = self.get_exa_by_generator()
+
+        xwalk = self.xwalk
+
+        df_860 = self.get_860_by_x(subplant_id_col="generator_id")
+
+        # list of cols we need for melt
+
+        """
+        missing data type #1: find generators that don't repeat for complete year range
+        and spit out plant / gen / missing year / prime / fuel
+
+        prime and fuel based on latest reported (and highest mmbtu) PF
+        observation in allocated GF923
+
+        """
+
+        missing_years = (
+            hist_data.assign(
+                n_years_gen=lambda x: x.groupby(["plant_id_eia", "generator_id"])[
+                    "report_date"
+                ].transform("nunique"),
+                report_year=lambda x: x.report_date.dt.year,
+            )
+            .query("n_years_gen != 15")
+            .groupby(["plant_id_eia", "generator_id"])["report_year"]
+            .agg(list)
+            .reset_index()
+            .assign(
+                needed_years=lambda x: x.apply(
+                    lambda _: [*range(2006, 2020, 1)], axis=1
+                ),
+                missing_years=lambda x: (
+                    x["needed_years"].map(set) - x["report_year"].map(set)
+                ),
+                years_to_fill_in=lambda x: x["missing_years"].apply(list),
+            )
+        )
+
+        # create a row for each year that needs to be filled in
+        split = pd.DataFrame([pd.Series(x) for x in missing_years.years_to_fill_in])
+        split.columns = [f"year_{x + 1}" for x in split.columns]
+
+        missing_years_clean = pd.concat([missing_years, split], axis=1)
+
+        missing_years_clean = (
+            missing_years_clean.melt(
+                id_vars=["plant_id_eia", "generator_id"],
+                value_vars=[
+                    "year_1",
+                    "year_2",
+                    "year_3",
+                    "year_4",
+                    "year_5",
+                    "year_6",
+                    "year_7",
+                    "year_8",
+                    "year_9",
+                    "year_10",
+                    "year_11",
+                    "year_12",
+                    "year_13",
+                    "year_14",
+                ],
+                value_name="year",
+            )
+            .query("year.notnull()")
+            .assign(
+                year=lambda x: x["year"].astype("Int64"), fuss=lambda x: "missing_years"
+            )
+            .merge(
+                self.get_gf923_by_generator(counterfactuals=True)
```

Comment:
since this takes a while and is used twice, consider running this method and storing the result like `df_860`